lib/tasks/version_scheme_counter.rake

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+#
+# [
+#   ["Maven", "io.redlink.geocoding:proxy-commons"],
+#   ["Packagist", "phraseanet/php-sdk"]
+# ].each { |project| PackageManager.const_get(project[0]).update project[1]}
+#
+#
+# Version scheme detection methodology:
+# 1. Test against all schemes and keep a tally of which ones validate.
+# 2. The one with the highest tally is likely the scheme which was being followed.
+# 3. If tied for unknown, mark unknown.
+# 4. Since calver validates as other schemes, prefer calver in the case of a tie since it is unusual and likely calver.
+# 5. In the case of a tie, if Semver is one of the contenders, prefer Semver for its ubiquity.
+# 6. If a tie between Maven and PEP440, check platform and choose accordingly.
+#
+# Known issues: many version schemes will match PEP440 and even Maven and yet be Semver. In the case of a tie, we choose
+#   semver, but this may turn out to be the wrong choice later on.
+
+module VersionSchemeDetection
+  REGEXES = {
+    MAVEN: /^(0|([1-9]\d*))(\.[1-9]\d*){1,2}(-(((alpha|beta|rc)-)?\d+|SNAPSHOT))?$/,
+    # PEP440 Regex copied from https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions
+    PEP440: /^([1-9][0-9]*!)?(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*((a|b|rc)(0|[1-9][0-9]*))?(\.post(0|[1-9][0-9]*))?(\.dev(0|[1-9][0-9]*))?$/
+  }
+
+  # validators return truthy/falsy
+  VALIDATORS = {
+    MAVEN: ->(version) { version =~ REGEXES[:MAVEN] },
+    PEP440: ->(version) { version =~ REGEXES[:PEP440] },
+    SEMVER: ->(version) { SemanticRange.valid(version).present? },
+    CALVER: lambda do |version|
+      first_part = version.split(".")[0]
+      # Assume most if not all Calver versions were released after 1000 AD
+      # Also assume if they are using YY as the first part that it is after 2000 AD
+      first_part.length >= 4 || (first_part.length == 2 && first_part.to_i > 0)
+    end
+  }
+end
+
+tallies = { semver: 0, pep440: 0, maven: 0, calver: 0, unknown: 0, no_versions: 0 }
+
+namespace :version do
+  desc 'Tests a sampling of project\'s versions to count occurrences of different versioning schemes'
+  task scheme_counter: :environment do
+    # projects = Project.order(Arel.sql("RANDOM()")).limit(100)
+
+    global_tallies = tallies.clone.merge({
+      unknown_versions: []
+    })
+
+    Project.find_each(batch_size: 1000) do |project|
+      local_tallies = tallies.clone
+
+      if !project.versions_count?
+        global_tallies[:no_versions] += 1
+        next
+      end
+
+      project.versions.each do |version|
+        version_number = version.number
+        matchable = false
+        [
+          :semver,
+          :pep440,
+          :maven,
+          :calver,
+        ].each do |scheme|
+          if VersionSchemeDetection::VALIDATORS[scheme.upcase].call(version_number)
+            local_tallies[scheme] += 1
+            matchable = true
+          end
+        end
```

Comment:
Just to 2x check, this doesn't break on purpose so that we catch all matching schemes for a version? Makes sense, to get the best approximation

Reply:
yep, that's right